### PR TITLE
Add `sdk-client/middleware-correlation-id`

### LINKED ISSUE #551: Add support for  a `correlationId` on each request to `sdk-client`
### Description

This commercetools platform supports specifying correlation ids on each request. Whenever no correlation id is specified the platform will generate one for the incoming request right after it hit the API. This correlation id is then passed through the system (like correlation ids are) for request tracing at different stages. 

Other SDKs already support specifying a custom correlation id. The id is then passed as a request header `x-correlation-id` (string with max 256 characters). The commercetools API then takes the specified id over generating a new one. This allows clients to "bridge" request tracing from an external system into our APIs. 

Correlation ids are becoming needed for the merchant center to:

1. Trace failing requests from e.g. Sentry in internal commercetools logging tools
2. Make known that a specific request comes from the merchant center (as prefix)
3. Allow for anonymous user tracing/identification to be aware which user modified data in the merchant center

From these the requirements are:

1. Add a `sdk-middleware-correlation-id` (suggestion)
2. Allow configuration and prefixing of the correlation id
3. Send the generated correlation id from the middleware as a header

### PR DESCRIPTION
**Pull request by tdeekens**
#### Summary

Closes #551 

#### Description

This adds the correlation id middleware discussed in the issue. It's not necessarily meant to be perfect yet. However, I felt some code would help to continue the discussion. This in my opinion would be the middleware in its simplest form.

### CODE REVIEW

File: docs/sdk/api/sdkMiddlewareCorrelationId.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# `sdk-middleware-logger`
```
Comment (emmenko) [inline docs/sdk/api/sdkMiddlewareCorrelationId.md]: Typo

File: docs/sdk/api/sdkMiddlewareCorrelationId.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#### Node.js
+
+```bash
+npm install --save @commercetools/sdk-middleware-correlation-id
+```
+
+#### Browser
+
+```html
+<script src="https://unpkg.com/@commercetools/sdk-middleware-logger/dist/commercetools-sdk-middleware-correlation-id.umd.min.js"></script>
```
Comment (emmenko) [inline docs/sdk/api/sdkMiddlewareCorrelationId.md]: Typo

Comment (tdeekens) [reply docs/sdk/api/sdkMiddlewareCorrelationId.md]: WAAAAT. I grepped them like a manic! :)

File: docs/sdk/api/sdkMiddlewareCorrelationId.md (hunk 1 above)
Comment (tdeekens) [reply docs/sdk/api/sdkMiddlewareCorrelationId.md]: Thanks. Fixed.

**Comment by tdeekens** [comment]
Note in 7c53d9b I fixed a minor config glitch I noticed during build outputs.

File: docs/sdk/api/README.md [hunk 3]
```diff
@@ -47,41 +48,54 @@ Middelware to authenticate the [request](/sdk/Glossary.md#clientrequest) using o
...
 * [getErrorByCode(code)](/sdk/api/sdkMiddlewareHttp.md#geterrorbycode)
 
 ### `sdk-middleware-queue`
+
 Middelware to throttle concurrent [request](/sdk/Glossary.md#clientrequest) to a certain limit. Useful to reduce concurrent HTTP requests.
 
 * [createQueueMiddleware(options)](/sdk/api/sdkMiddlewareQueue.md#createqueuemiddlewareoptions)
 
 ### `sdk-middleware-logger`
+
```
Comment (daern91) [inline docs/sdk/api/README.md]: another old typo here... if you wanna fix it

File: docs/sdk/api/README.md [hunk 4]
```diff
@@ -47,41 +48,54 @@ Middelware to authenticate the [request](/sdk/Glossary.md#clientrequest) using o
...
 Middelware to log incoming [request](/sdk/Glossary.md#clientrequest) and [response](/sdk/Glossary.md#clientrequest) objects.
 
 * [createLoggerMiddleware(options)](/sdk/api/sdkMiddlewareLogger.md#createloggermiddlewareoptions)
 
 ### `sdk-middleware-user-agent`
+
 Middelware to automatically set the `User-Agent` to the [request](/sdk/Glossary.md#clientrequest).
```
Comment (daern91) [inline docs/sdk/api/README.md]: same

File: docs/sdk/api/sdkMiddlewareCorrelationId.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# `sdk-middleware-correlation-id`
+
+Middelware add a correlation id to [requests](/sdk/Glossary.md#clientrequest).
```
Comment (daern91) [inline docs/sdk/api/sdkMiddlewareCorrelationId.md]: typo

Comment (tdeekens) [reply docs/sdk/api/sdkMiddlewareCorrelationId.md]: Did I get em all now? :)

File: packages/sdk-middleware-correlation-id/package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{
+  "private": false,
+  "publishConfig": {
+    "access": "public"
+  },
+  "name": "@commercetools/sdk-middleware-correlation-id",
+  "version": "1.0.5",
```
Comment (wizzy25) [inline packages/sdk-middleware-correlation-id/package.json]: How did we get to `v1.0.5`? 😳

Comment (tdeekens) [reply packages/sdk-middleware-correlation-id/package.json]: Sorry. Copied over will change.

**Comment by tdeekens** [comment]
No. The middleare merely offers a hook into the system to be asked unqiue strings as a correlation id. So eg the MC will use uuid4. The generate is almost the return value of uuid4 with some prefixing. There is really not much to it apart that the uuid will be set as the header. This means that requests are traceable starting with the sdk.

**Comment by tdeekens** [comment]
I tested this linked into the MC. Works fine. I can walk you through the details of why/how later. I'd like to continue with this @wizzy25.

**Pull request merged**